File: cli/src/core/references/sources/NotionEnvelope.test.ts
import { describe, expect, it } from "vitest";
import { parseNotionEnvelope } from "./NotionEnvelope.js";

describe("parseNotionEnvelope", () => {
	it("extracts the body inside <content>…</content>", () => {
		const text = `<page>
  <title>Sample</title>
  <metadata>type=page</metadata>
  <content># Heading

Body of the page with **markdown**.</content>
</page>`;
		const out = parseNotionEnvelope(text);
		expect(out.content).toBe("# Heading\n\nBody of the page with **markdown**.");
	});

	it("preserves newlines inside the content (multi-line dotall match)", () => {
		const text = "<content>line one\nline two\nline three</content>";
		expect(parseNotionEnvelope(text).content).toBe("line one\nline two\nline three");
	});

	it("returns empty content for input with no <content> block", () => {
		expect(parseNotionEnvelope("<page><title>x</title></page>").content).toBe("");
	});

	it("returns empty content for malformed input (only opening tag, no close)", () => {
		expect(parseNotionEnvelope("<content>oops no close tag").content).toBe("");
	});

	it("returns empty content for an empty string input", () => {
		expect(parseNotionEnvelope("").content).toBe("");
	});

	it("returns empty content for a non-string input (defensive — runtime safety)", () => {
		// Cast through unknown to satisfy the type system; we still want runtime
		// guard behaviour to be observable.
		expect(parseNotionEnvelope(undefined as unknown as string).content).toBe("");
		expect(parseNotionEnvelope(null as unknown as string).content).toBe("");
	});

	it("captures only the first <content> block when multiple appear", () => {
		const text = "<content>first</content><content>second</content>";
		expect(parseNotionEnvelope(text).content).toBe("first");
	});

	it("captures an empty body when <content></content> is empty", () => {
		expect(parseNotionEnvelope("<content></content>").content).toBe("");
	});

	it("extracts content when the <content> open tag carries attributes", () => {
		// notion-fetch may evolve to emit `<content type="markdown">…`. Without
		// attribute tolerance the body would be silently dropped (returns "").
		expect(parseNotionEnvelope('<content type="markdown">body here</content>').content).toBe("body here");
	});
});

File: intellij/src/test/kotlin/ai/jolli/jollimemory/core/references/NotionEnvelopeTest.kt
package ai.jolli.jollimemory.core.references

import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.Test

class NotionEnvelopeTest {

	@Test
	fun `extracts content block`() {
		val text = "<page><title>Hello</title><content># Body\n\nSome text</content></page>"
		NotionEnvelope.parse(text).content shouldBe "# Body\n\nSome text"
	}

	@Test
	fun `returns empty for missing content block`() {
		NotionEnvelope.parse("<page><title>Hello</title></page>").content shouldBe ""
	}

	@Test
	fun `returns empty for null input`() {
		NotionEnvelope.parse(null).content shouldBe ""
	}

	@Test
	fun `returns empty for empty string`() {
		NotionEnvelope.parse("").content shouldBe ""
	}

	@Test
	fun `tolerates attributes on content tag`() {
		val text = """<content class="body">inner</content>"""
		NotionEnvelope.parse(text).content shouldBe "inner"
	}

	@Test
	fun `takes first content block only`() {
		val text = "<content>first</content><content>second</content>"
		NotionEnvelope.parse(text).content shouldBe "first"
	}
}
